aiida/sphinxext/process.py

```diff
@@ -125,45 +125,57 @@ def build_portnamespace_doctree(self, port_namespace):
         """
         from aiida.engine.processes.ports import InputPort, PortNamespace
 
+        if not port_namespace:
+            return
         result = nodes.bullet_list(bullet='*')
         for name, port in sorted(port_namespace.items()):
             item = nodes.list_item()
             if _is_non_db(port) and self.HIDE_UNSTORED_INPUTS_FLAG in self.options:
                 continue
             if isinstance(port, (InputPort, OutputPort)):
-                item += self.build_port_paragraph(name, port)
+                item.extend(self.build_port_content(name, port))
             elif isinstance(port, PortNamespace):
                 item += addnodes.literal_strong(text=name)
                 item += nodes.Text(', ')
                 item += nodes.emphasis(text='Namespace')
                 if port.help is not None:
                     item += nodes.Text(' -- ')
                     item.extend(publish_doctree(port.help)[0].children)
-                item += self.build_portnamespace_doctree(port)
+                sub_doctree = self.build_portnamespace_doctree(port)
+                if sub_doctree:
+                    # This is a workaround because this extension doesn't work with Python2.
+                    try:
+                        from sphinxcontrib.details.directive import details, summary
+                        sub_item = details()
+                        sub_item += summary(text='Namespace Ports')
+                        sub_item += sub_doctree
+                    except ImportError:
+                        sub_item = sub_doctree
+                    item += sub_item
             else:
                 raise NotImplementedError
             result += item
         return result
 
-    def build_port_paragraph(self, name, port):
+    def build_port_content(self, name, port):
         """
-        Build the paragraph that describes a single port.
+        Build the content that describes a single port.
         """
-        paragraph = nodes.paragraph()
-        paragraph += addnodes.literal_strong(text=name)
-        paragraph += nodes.Text(', ')
-        paragraph += nodes.emphasis(text=self.format_valid_types(port.valid_type))
-        paragraph += nodes.Text(', ')
-        paragraph += nodes.Text('required' if port.required else 'optional')
+        res = []
```

Comment:
I guess this switch from `paragraph` was necessary to make the spacing consistent?
Perhaps a comment to explain this might be useful.

Reply:
Yeah, that's exactly why. I did document that in the commit message, but since the paragraph is no longer there in the code I'm not sure a code comment would really be understandable.